app/models/katello/product.rb

```diff
@@ -253,5 +256,18 @@ def self.unused_product_id
     class Jail < ::Safemode::Jail
       allow :name, :label
     end
+
+    private
+
+    def ensure_provider_type_matches_id
+      if provider&.redhat_provider? && ::Katello::Glue::Candlepin::Product.custom_product_id?(cp_id)
```

Comment:
when can provider be nil?

Reply:
I actually hit a NilClass error here, but I don't remember exactly where. It may have been in a test.

Reply:
Definitely work looking into. The code communicates that provider can be nil. If that's true,  then 👍 . But if that's not true then it's confusing (and wrong)!

Reply:
It was in a test where `provider` wasn't assigned. It was just easier to add one character than it is to update a bunch of test fixtures.

Reply:
oh no I typed up a long comment here and now I lost it, let me know if you don't see it!

Reply:
Found it: https://github.com/Katello/katello/commit/792f3b464d57e81c650d549686f24d6e6070a831#r101064846